orchestrator/cosmos/broadcast.go

```diff
@@ -333,16 +334,53 @@ func (s *peggyBroadcastClient) sendValsetUpdateClaims(
 	return nil
 }
 
+func (s *peggyBroadcastClient) sendERC20DeployedClaims(
+	ctx context.Context,
+	erc20Deployed *wrappers.PeggyERC20DeployedEvent,
+) error {
+
+	log.WithFields(log.Fields{
+		"nonce":          erc20Deployed.EventNonce.String(),
+		"token_contract": erc20Deployed.TokenContract.Hex(),
+		"cosmos_denom":   erc20Deployed.CosmosDenom,
+		"event_nonce":    erc20Deployed.EventNonce.String(),
+	}).Infoln("Oracle observed an ERC20 deployed event. Sending MsgERC20DeployedClaim")
+
+	msg := &types.MsgERC20DeployedClaim{
+		EventNonce:    erc20Deployed.EventNonce.Uint64(),
+		BlockHeight:   erc20Deployed.Raw.BlockNumber,
+		Orchestrator:  s.AccFromAddress().String(),
+		CosmosDenom:   erc20Deployed.CosmosDenom,
+		TokenContract: erc20Deployed.TokenContract.Hex(),
+		Name:          erc20Deployed.Name,
+		Decimals:      uint64(erc20Deployed.Decimals),
+		Symbol:        erc20Deployed.Symbol,
+	}
+
+	if txResponse, err := s.broadcastClient.SyncBroadcastMsg(msg); err != nil {
+		log.WithError(err).Errorln("broadcasting ERC20DeployedClaim failed")
+		return err
+	} else {
+		log.WithFields(log.Fields{
+			"event_nonce": erc20Deployed.EventNonce.String(),
+			"txHash":      txResponse.TxHash,
+		}).Infoln("Oracle sent ERC20Deployed event succesfully")
+	}
+
+	return nil
+}
+
 func (s *peggyBroadcastClient) SendEthereumClaims(
 	ctx context.Context,
 	lastClaimEvent uint64,
 	deposits []*wrappers.PeggySendToCosmosEvent,
 	withdraws []*wrappers.PeggyTransactionBatchExecutedEvent,
 	valsetUpdates []*wrappers.PeggyValsetUpdatedEvent,
+	erc20Deployed []*wrappers.PeggyERC20DeployedEvent,
 ) error {
 
-	totalClaimEvents := len(deposits) + len(withdraws) + len(valsetUpdates)
-	var count, i, j, k int
+	totalClaimEvents := len(deposits) + len(withdraws) + len(valsetUpdates) + len(erc20Deployed)
```

Comment:
not meant to block here, but in the future we gotta think of a cleaner way to handling this logic. This is pretty difficult to read. LMK your thoughts on how we can better generalize this code.

Reply:
Took me a while to figure out what was going on here 😅 My first thought is to get them all into a single list ordered by the event nonce. Maybe using some kind of wrapper so we don't have to use casting. Something like this:

```go
	// SortableEvent exists with the only purpose to make a nicer sortable slice
	type SortableEvent struct {
		EventNonce         uint64
		DepositEvent       *wrappers.PeggySendToCosmosEvent
		WithdrawEvent      *wrappers.PeggyTransactionBatchExecutedEvent
		ValsetUpdateEvent  *wrappers.PeggyValsetUpdatedEvent
		ERC20DeployedEvent *wrappers.PeggyERC20DeployedEvent
	}
	allevents := []SortableEvent{}
	// add all events to allevents
	// sort allevents by event nonce
	// remove all events which nonce is equal or lower than lastClaimEvent
	for _, event := range allevents {
		if event.DepositEvent != nil {
			if err := s.sendDepositClaims(ctx, event.DepositEvent); err != nil {
				log.WithError(err).Errorln("broadcasting MsgDepositClaim failed")
				return err
			}
		}

		if event.WithdrawEvent != nil {
			if err := s.sendWithdrawClaims(ctx, event.WithdrawEvent); err != nil {
				log.WithError(err).Errorln("broadcasting MsgWithdrawClaim failed")
				return err
			}
		}

		// do the same for valsets
		// do the same for erc20 deployed
	}
```

Reply:
I like it! Can you create an issue with a proposal?